File: core/src/main/resources/org/treblereel/gwt/three4g/cameras/Camera.js
import { Matrix4 } from '../math/Matrix4.js';
import { Object3D } from '../core/Object3D.js';
import { Vector3 } from '../math/Vector3.js';

class Camera extends Object3D {

	constructor() {

		super();

		/** @type {string} */
		this.type = 'Camera';
		/** @type {Matrix4} */
		this.matrixWorldInverse = new Matrix4();
		/** @type {Matrix4} */
		this.projectionMatrix = new Matrix4();
		/** @type {Matrix4} */
		this.projectionMatrixInverse = new Matrix4();
		/** @type {boolean} */
		this.isCamera = true;

	}

	/**
     * @suppress {checkTypes} 
	 * @param {Camera} source 
	 * @param {boolean=} recursive 
	 * @return {Camera}
	 */
	copy( source, recursive ) {

		Object3D.prototype.copy.call( this, source, recursive );

		this.matrixWorldInverse.copy( source.matrixWorldInverse );

		this.projectionMatrix.copy( source.projectionMatrix );
		this.projectionMatrixInverse.copy( source.projectionMatrixInverse );

		return this;

	}

	/**
	 * @param {Vector3=} target 
	 */
	getWorldDirection( target ) {

		if ( target === undefined ) {

			console.warn( 'THREE.Camera: .getWorldDirection() target is now required' );
			target = new Vector3();

		}

		this.updateWorldMatrix( true, false );

		const e = this.matrixWorld.elements;

		return target.set( - e[ 8 ], - e[ 9 ], - e[ 10 ] ).normalize();

	}

	/**
	 * @param {boolean=} force 
	 */
	updateMatrixWorld( force ) {

		Object3D.prototype.updateMatrixWorld.call( this, force );

		this.matrixWorldInverse.copy( this.matrixWorld ).invert();

	}

	/**
	 * 
	 * @param {boolean} updateParents 
	 * @param {boolean} updateChildren 
	 */
	updateWorldMatrix( updateParents, updateChildren ) {

		Object3D.prototype.updateWorldMatrix.call( this, updateParents, updateChildren );

		this.matrixWorldInverse.copy( this.matrixWorld ).invert();

	}

	/**
	 * @return {Camera}
	 */
	clone() {

		return new this.constructor().copy( this );

	}

}

export { Camera };

File: core/src/main/java/org/treblereel/gwt/three4g/cameras/Camera.java
package org.treblereel.gwt.three4g.cameras;

import jsinterop.annotations.JsPackage;
import jsinterop.annotations.JsType;
import org.treblereel.gwt.three4g.math.Vector3;
import org.treblereel.gwt.three4g.core.Object3D;
import org.treblereel.gwt.three4g.math.Matrix4;

@JsType(isNative = true, name = "THREE.Camera", namespace = JsPackage.GLOBAL)
public class Camera extends Object3D {
  public boolean isCamera;
  public Matrix4 matrixWorldInverse;
  public Matrix4 projectionMatrix;
  public Matrix4 projectionMatrixInverse;

  public native Vector3 getWorldDirection(Vector3 target);

  public native void updateMatrixWorld();

  public native void updateMatrixWorld(boolean force);
}
